### PR TITLE
Add data-plane pre upgrade checks

### PR DESCRIPTION
**Pull request by gabrielcocenza**
- check if control-plane is already upgraded if user pass data-pane upgrade group
- check if data-plane exists if user pass data-pane upgrade group
- splitted the logic on determine_upgrade_target to several functions
- removed parameters of cli functions module by passing args object -

### CODE REVIEW

File: cou/steps/plan.py [hunk 1]
```diff
@@ -52,24 +53,163 @@
...
+
+    :param analysis_result: Analysis result.
+    :type analysis_result: Analysis
+    :raises NoTargetError: When cannot find target to upgrade
+    :raises OutOfSupportRange: When the upgrade scope is not supported
+        by the current series.
+    """
+    current_os_release = analysis_result.current_cloud_os_release
+    current_series = analysis_result.current_cloud_series
+    if current_os_release and current_series:
```
Comment (rgildein) [inline cou/steps/plan.py:145]: We do not need to check this or do we? It was checked in `is_valid_openstack_cloud`.

File: cou/steps/plan.py [hunk 2]
```diff
@@ -52,24 +53,163 @@
...
+    :type analysis_result: Analysis
+    :return: Whether the control plane is already upgraded or not.
+    :rtype: bool
+    """
+    control_plane = analysis_result.min_os_version_control_plane
+    data_plane = analysis_result.min_os_version_data_plane
+
+    if control_plane and data_plane:
+        return control_plane > data_plane
+    return False
```
Comment (rgildein) [inline cou/steps/plan.py:198]:
```suggestion
    return control_plane and data_plane and control_plane > data_plane
```

File: cou/steps/plan.py [hunk 3]
```diff
@@ -52,24 +53,163 @@
...
+    """Check the Ubuntu series of the cloud to see if it is supported.
+
+    :param analysis_result: Analysis result.
+    :type analysis_result: Analysis
+    :raises OutOfSupportRange: When series is not supported.
+    """
+    current_series = analysis_result.current_cloud_series
+    supporting_lts_series = ", ".join(LTS_TO_OS_RELEASE)
+    # series already checked at is_valid_openstack_cloud
+    if current_series not in supporting_lts_series:  # type: ignore
```
Comment (rgildein) [inline cou/steps/plan.py:109]: I think it's cleaner to add `if current_series and current_series not in ...`, this check should not raise an exception if `current_series` is `None`.

Comment (rgildein) [reply cou/steps/plan.py:109]: same for all other `# type: ignore`

File: cou/cli.py [hunk 4]
```diff
@@ -107,19 +110,15 @@ async def continue_upgrade() -> bool:
     return False
 
 
-async def analyze_and_plan(
-    model_name: Optional[str], backup_database: bool
-) -> tuple[Analysis, UpgradePlan]:
+async def analyze_and_plan(args: argparse.Namespace) -> tuple[Analysis, UpgradePlan]:
```
[the remark this answers is not in the document]
Comment (gabrielcocenza) [reply cou/cli.py:113]: I would need to add a new argument here for the `upgrade_group` on some functions and IMO they are having too many arguments and as reference by the clean code book:

> The ideal number of arguments for a function is zero (niladic). Next comes one, followed close by two. Three arguments should be avoided where possible. More than three requires very special justification and then shouldn't be used anyway.

All the arguments that we need for those functions are on args, so why not pass it and have shorter description and arguments in the function

File: cou/cli.py [hunk 5]
```diff
@@ -130,22 +129,24 @@ async def analyze_and_plan(
     logger.info(analysis_result)
     progress_indicator.succeed()
 
+    progress_indicator.start("Making sane checks to upgrade...")
+    pre_plan_sane_checks(args.upgrade_group, analysis_result)
```
[the remark this answers is not in the document]
Comment (gabrielcocenza) [reply cou/cli.py:133]: This is the group passed by the user. Ex:
`cou upgrade data-plane` results in `upgrade_group` = `data-plane`

[the remark this answers is not in the document]
Comment (rgildein) [reply cou/cli.py:133]: Ach I see so the `upgrade_groub` cloud be `data-plane` or `control-plane` only.

File: cou/steps/plan.py [hunk 6]
```diff
@@ -52,24 +53,163 @@
...
+    :param analysis_result: Analysis result.
+    :type analysis_result: Analysis
+    """
+    checks = [
+        is_valid_openstack_cloud,
+        is_supported_series,
+        is_highest_release_achieved,
+        is_target_supported,
+    ]
+    if upgrade_group == "data-plane":
```
[the remark this answers is not in the document]
Comment (rgildein) [reply cou/steps/plan.py:70]: Now I understand that is not good idea doing as flag, but at least we should use `"data-plane"` as constant, so if we changed those group we can do it in single place.

File: cou/cli.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gabrielcocenza) [reply cou/cli.py:113]: If we test properly the generation of the object `Namespace`, then we are covered. I don't think it's out of context. The book this session:

> When a function seems to need more than two or three arguments, it is likely that some of those arguments ought to be wrapped into a class of their own. Reducing the number of arguments by creating objects out of them may seem like cheating, but it's not.

[the remark this answers is not in the document]
Comment (rgildein) [reply cou/cli.py:113]: > wrapped into a class of their own

You are not wrapping it to a specific class, you are using Namespace, and that is the same as using dict.

**Comment by gabrielcocenza** [comment]
PR moved to #222

**Pull request closed without merging**